Fix credential method selection for chained vs non-chained MFA

propilideno opened the pull request:
## Introduction

After my changes, I notice that I broke some features when I was trying to pass the session duration to my `PROFILENAME`.

```
$ avault76 login PROFILENAME -d 8h --debug                    
2025/08/16 16:37:20 aws-vault v7.6.0-rc1
2025/08/16 16:37:20 Using prompt driver: terminal
2025/08/16 16:37:20 [keyring] Considering backends: [secret-service]
2025/08/16 16:37:20 Loading config file /home/***/.aws/config
2025/08/16 16:37:20 Parsing config file /home/***/.aws/config
2025/08/16 16:37:20 profile PROFILENAME: using stored credentials
2025/08/16 16:37:20 profile PROFILENAME: skipping GetSessionToken because sessions are disabled for this profile
2025/08/16 16:37:20 profile PROFILENAME: using GetSessionToken (with MFA)
2025/08/16 16:37:20 profile PROFILENAME: using AssumeRole (chained MFA)
2025/08/16 16:37:20 Re-using cached credentials ****************ICW7 from sts.GetSessionToken, expires in 7h12m54.65693466s
aws-vault: error: login: operation error STS: AssumeRole, https response error StatusCode: 400, RequestID: 2d939bc8-4d67-4019-a676-2b87ed6e3bf8, api error ValidationError: The requested DurationSeconds exceeds the 1 hour session limit for roles assumed by role chaining.
```
In this example, you can notice that we are dealing it as a `(chained MFA)`, that is not the truth.

## Fixed
This problem occurs because the: https://github.com/ByteNess/aws-vault/blob/7500108301cf80f75e008d809eb9e9654e18d40e/vault/vault.go#L344-L346

Notice that previously I was handling both methods as **chained credentials**, so to use this feature `--duration`, we must use the right method at the right time

I could get it working in both cases by using `NewSessionTokenProvider` in the **SOURCE** profile when we have `config.ChainedFromProfile != nil`

## Cached Credentials

https://github.com/ByteNess/aws-vault/blob/7500108301cf80f75e008d809eb9e9654e18d40e/vault/sessionkeyring.go#L103-L114

After debugging, I found it:
```
 key github.com/99designs/aws-vault/v7/vault.SessionMetadata = vault.SessionMetadata {Type: "sts.GetSessionToken", ProfileName: "PROFILENAME", MfaSerial: "arn:aws:iam::***:mfa/***", Expiration: time.Time(0001-01-01T00:00:00Z){wall: 0, ext: 0, loc: *time.Location nil}}
   Type string = "sts.GetSessionToken"
   ProfileName string = "PROFILENAME"
   MfaSerial string = "arn:aws:iam::***:mfa/***"
  Expiration time.Time = time.Time(0001-01-01T00:00:00Z){wall: 0, ext: 0, loc: *time.Location nil}
    wall uint64 = 0 = 0x0
    ext int64 = 0
    loc *time.Location = *time.Location nil
```

```
strings.HasPrefix
  keyName:                 sts.AssumeRole,YWNr,YXJuOmF3czppYW06OjM4MDg2NTgzMDYxNTptZmEZQ,1755390794
  key.StringForMatching(): sts.GetSessionToken,YWNr,YXJuOmF3czppYW06OjM4MDg2NTgzMDYxNTptZmEZQ,
```

Notice that we only reuse if we use the same sts get credential method,
I tried to bypass it but got a **403 Forbidden**. So I'll leave it unchanged, because i cannot find a way to do it better


## Conclusion

aws-vault after a login without role chaining

```
Enter MFA code for arn:aws:iam::***:mfa/***: 
$ aws-vault list

Profile                          Credentials                      Sessions                                        
=======                          ===========                      ========                                        
PROFILENAME                      PROFILENAME                      sts.GetSessionToken:7h59m59s
```

aws-vault after a login with role chaining

```
$ aws-vault list

Profile                          Credentials                      Sessions                                        
=======                          ===========                      ========                                        
PROFILENAME                      PROFILENAME                      sts.AssumeRole:59m59s
```

aws-vault after both
```
$ aws-vault list

Profile                          Credentials                      Sessions                                        
=======                          ===========                      ========                                        
PROFILENAME                      PROFILENAME                      sts.GetSessionToken:7h59m18s, sts.AssumeRole:59m52s
```

So, we must provide `Enter MFA code for arn:aws:iam::***:mfa/***:` for each method.

propilideno commented:
> Thanks @propilideno ! I'll review this shorty, but would you mind opening a PR from a for of this repo instead of the upstream - I think that's why PR checks might be failing :)

Hello @mbevc1, it's failing because I used `merge:` as the prefix of my commit message.
I did the merge from ByteNess/aws-vault. Because of that, I closed #86, which was the previous one that was failing due to merge conflicts. So, I created a branch named `main`; the previous one was based on `master`. I don't want to mix them because I will leave the PR to 99designs open.

mbevc1 commented:
It's checking PR title, which I've fixed. But I think failure now might be related to integration access. I was wondering if we could test this with the for from here and now upstream. Mind, not a blocker for review of the change.

mbevc1 commented:
Yeah, I've already changed that but I reckon using other forks might break this :thinking: . Would you be able to give it a go to test this out?

propilideno commented:
It looks working for me: https://github.com/propilideno/aws-vault/pull/1

mbevc1 commented:
So, back to the PR - this is now breaking source profile chaining where it keeps asking for MFA and not assuming the role:
```
Enter MFA code for arn:aws:iam::xxxxxxxxxxxxx:mfa/<user>: XXXXXX
2025/08/17 22:20:34 Re-using cached credentials ******************** from sts.GetSessionToken, expires in 35m20.33138408s
aws-vault: error: login: operation error STS: AssumeRole, get identity: get credentials: operation error STS: GetSessionToken, https response error StatusCode: 403, RequestID: 7fdc63fe-085c-410d-adee-312816d6b964, api error AccessDenied: Cannot call GetSessionToken with session credentials
```

mbevc1 commented on vault/vault.go:
```diff
@@ -254,22 +254,9 @@ func (t *TempCredentialsCreator) getSourceCredWithSession(config *ProfileConfig,
+	isChainedCredentialSource := config.ChainedFromProfile != nil
 
-	if config.HasRole() {
+	if config.HasRole() && !isChainedCredentialSource {
```
Is this correct logic here checking for not chained cred source? :thinking:

propilideno replied:
solved on 1912e2689df22de712d9843de65579414cbddebc

propilideno commented:
> So, back to the PR - this is now breaking source profile chaining where it keeps asking for MFA and not assuming the role:
> 
> ```
> Enter MFA code for arn:aws:iam::xxxxxxxxxxxxx:mfa/<user>: XXXXXX
> 2025/08/17 22:20:34 Re-using cached credentials ******************** from sts.GetSessionToken, expires in 35m20.33138408s
> aws-vault: error: login: operation error STS: AssumeRole, get identity: get credentials: operation error STS: GetSessionToken, https response error StatusCode: 403, RequestID: 7fdc63fe-085c-410d-adee-312816d6b964, api error AccessDenied: Cannot call GetSessionToken with session credentials
> ```

Oh, you are right. I do not use this feature, so i couldn't notice it. I also runned unit tests but all cases it passed.

mbevc1 commented:
No worries, thanks!

propilideno commented:
Hello @mbevc1, sorry for the slow response. I didn’t have time to work on my open source contributions this month.

I tested some scenarios, and everything below is working for me:
- [x] source chaining
- [x] role chaining
- [x] assume role with default duration
- [x] assume role with high duration

I'll wait for your tests to confirm if the issue is resolved.

Best Regards,
Lucas Almeida.

propilideno commented:
Hi @mbevc1,

Sorry for the late response, and thank you for your patience.

I revisited this PR and pushed a focused update to address the issue you reported (source-profile chaining being treated like role chaining), while preserving the original role-chaining MFA fix from #75.

I updated `getSourceCredWithSession` to distinguish:

- profile chaining topology (`ChainedFromProfile`)
- real AWS role-chaining flow

With this, `GetSessionToken` is only used when needed to preserve MFA context for chained role assumptions, and non-chained role login is no longer treated as chained MFA flow.

I added test coverage with:

- `TestRoleProfileNotUsedAsRoleSourceSkipsGetSessionToken` (direct role login)
- `TestRoleChainingMfaPrimesSessionThenAssumes` (role -> role chaining with MFA)
- `TestSourceChainingNonRoleTargetDoesNotTreatAsRoleChaining` (role source -> non-role target)
- `TestNonRoleSourceProfileUsesGetSessionTokenWhenAllowed` (non-role source -> role target)
- `TestGetSessionTokenSkippedForDisabledSessionsNonRoleReturnsSourceCreds` (sessions-disabled non-role fallback)

#75 Running the unit tests
```text
--- FAIL: TestRoleProfileNotUsedAsRoleSourceSkipsGetSessionToken (0.00s)
    vault_test.go:177: did not expect GetSessionToken for non-chained role profile, logs:
        profile role1: using stored credentials
        profile role1: using GetSessionToken (with MFA)
        profile role1: using AssumeRole (chained MFA)
--- FAIL: TestSourceChainingNonRoleTargetDoesNotTreatAsRoleChaining (0.00s)
    vault_test.go:296: did not expect GetSessionToken for role source chained to non-role target, logs:
        profile leaf: sourcing credentials from profile role_source
        profile role_source: using stored credentials
        profile role_source: skipping GetSessionToken because profile 'leaf' has no MFA serial defined
        profile role_source: using GetSessionToken (with MFA)
        profile role_source: using AssumeRole (chained MFA)
        profile leaf: using GetSessionToken
```

#87 Running the unit tests
```text
--- PASS: TestRoleProfileNotUsedAsRoleSourceSkipsGetSessionToken (0.00s)
--- PASS: TestSourceChainingNonRoleTargetDoesNotTreatAsRoleChaining (0.00s)
```

I also validated with:

- `go test ./vault`
- `go test ./...`

And did a smoke test with my use-case scenarios and all of them worked.

mbevc1 commented:
Yes, please do that
